core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -460,25 +493,213 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+        } else if (isRunning) {
+          "running"
+        }
+      }
+
+      val launchTime = taskInfo.launchTime
+      val finishTime = if (!isRunning) taskInfo.finishTime else currentTime
+      val totalExecutionTime = finishTime - launchTime
+      minLaunchTime = launchTime.min(minLaunchTime)
+      maxFinishTime = launchTime.max(maxFinishTime)
+      numEffectiveTasks += 1
+
+      val metricsOpt = taskUIData.taskMetrics
+      val shuffleReadTime =
+        metricsOpt.flatMap(_.shuffleReadMetrics.map(_.fetchWaitTime)).getOrElse(0L)
+      val shuffleReadTimeProportion =
+        (shuffleReadTime.toDouble / totalExecutionTime * 100).toLong
+      val shuffleWriteTime =
+        (metricsOpt.flatMap(_.shuffleWriteMetrics
+          .map(_.shuffleWriteTime)).getOrElse(0L) / 1e6).toLong
+      val shuffleWriteTimeProportion =
+        (shuffleWriteTime.toDouble / totalExecutionTime * 100).toLong
+      val executorComputingTime = metricsOpt.map(_.executorRunTime).getOrElse(0L) -
+        shuffleReadTime - shuffleWriteTime
+      val executorComputingTimeProportion =
+        (executorComputingTime.toDouble / totalExecutionTime * 100).toLong
+      val serializationTime = metricsOpt.map(_.resultSerializationTime).getOrElse(0L)
+      val serializationTimeProportion =
+        (serializationTime.toDouble / totalExecutionTime * 100).toLong
+      val deserializationTime = metricsOpt.map(_.executorDeserializeTime).getOrElse(0L)
+      val deserializationTimeProportion =
+        (deserializationTime.toDouble / totalExecutionTime * 100).toLong
+      val gettingResultTime = getGettingResultTime(taskUIData.taskInfo)
+      val gettingResultTimeProportion =
+        (gettingResultTime.toDouble / totalExecutionTime * 100).toLong
+      val schedulerDelay = totalExecutionTime -
+        (executorComputingTime + shuffleReadTime + shuffleWriteTime +
+          serializationTime + deserializationTime + gettingResultTime)
+      val schedulerDelayProportion =
+        100 - executorComputingTimeProportion - shuffleReadTimeProportion -
+          shuffleWriteTimeProportion - serializationTimeProportion -
+          deserializationTimeProportion - gettingResultTimeProportion
+
+      val schedulerDelayProportionPos = 0
+      val deserializationTimeProportionPos =
+        schedulerDelayProportionPos + schedulerDelayProportion
+      val shuffleReadTimeProportionPos =
+        deserializationTimeProportionPos + deserializationTimeProportion
+      val executorRuntimeProportionPos =
+        shuffleReadTimeProportionPos + shuffleReadTimeProportion
+      val shuffleWriteTimeProportionPos =
+        executorRuntimeProportionPos + executorComputingTimeProportion
+      val serializationTimeProportionPos =
+        shuffleWriteTimeProportionPos + shuffleWriteTimeProportion
+      val gettingResultTimeProportionPos =
+        serializationTimeProportionPos + serializationTimeProportion
+
+      val timelineObject =
+        s"""
+           |{
```

Comment:
Do you need the "|"? They're used inconsistently it seems like, so can they just be eliminated?